Tribler/community/popularity/payload.py

```diff
@@ -1,48 +1,57 @@
 from __future__ import absolute_import
 
-from Tribler.pyipv8.ipv8.messaging.payload import Payload
-
-
-class ContentSubscription(Payload):
+import struct
 
-    format_list = ['I', '?']
+from Tribler.pyipv8.ipv8.messaging.payload import Payload
 
-    def __init__(self, identifier, subscribe):
-        super(ContentSubscription, self).__init__()
-        self.identifier = identifier
-        self.subscribe = subscribe
 
-    def to_pack_list(self):
-        data = [('I', self.identifier),
-                ('?', self.subscribe)]
-        return data
-
-    @classmethod
-    def from_unpack_list(cls, *args):
-        (identifier, subscribe) = args
-        return ContentSubscription(identifier, subscribe)
+TORRENT_INFO_FORMAT = '20sIIQ'  # Infohash, seeders, leechers and a timestamp
 
 
-class TorrentHealthPayload(Payload):
+class TorrentsHealthPayload(Payload):
 
-    format_list = ['20s', 'I', 'I', 'Q']
+    format_list = ['I', 'I', 'raw', 'raw']  # Number of random torrents, number of torrents checked by you
 
-    def __init__(self, infohash, num_seeders, num_leechers, timestamp):
-        super(TorrentHealthPayload, self).__init__()
-        self.infohash = infohash
-        self.num_seeders = num_seeders or 0
-        self.num_leechers = num_leechers or 0
-        self.timestamp = timestamp or 0
+    def __init__(self, random_torrents, torrents_checked):
+        """
+        Initialize a TorrentsHealthPayload, containing information on the health of both random torrents and popular
+        torrents that have been checked by you.
+        :param random_torrents: List of tuple of (infohash, seeders, leechers, checked_timestamp)
+        :param torrents_checked: List of tuple of (infohash, seeders, leechers, checked_timestamp)
+        """
+        super(TorrentsHealthPayload, self).__init__()
+        self.random_torrents = random_torrents
+        self.torrents_checked = torrents_checked
 
     def to_pack_list(self):
-        data = [('20s', self.infohash),
-                ('I', self.num_seeders),
-                ('I', self.num_leechers),
-                ('Q', self.timestamp)]
+        random_torrents_items = [item for sublist in self.random_torrents for item in sublist]
+        checked_torrents_items = [item for sublist in self.torrents_checked for item in sublist]
+        data = [('I', len(self.random_torrents)),
+                ('I', len(self.torrents_checked)),
+                ('raw', struct.pack("!" + TORRENT_INFO_FORMAT * len(self.random_torrents), *random_torrents_items)),
```

Comment:
This should be a `varlen` variant, `raw` can only appear once at the end of a message.

Reply:
Thanks! 👍 this is a bug indeed